File: annotation-processor/src/main/java/me/bristermitten/mittenlib/annotations/compile/ConfigImplGenerator.java
package me.bristermitten.mittenlib.annotations.compile;

import com.google.inject.Inject;
import com.squareup.javapoet.*;
import me.bristermitten.mittenlib.annotations.ast.ASTSettings;
import me.bristermitten.mittenlib.annotations.ast.AbstractConfigStructure;
import me.bristermitten.mittenlib.annotations.ast.ConfigTypeSource;
import me.bristermitten.mittenlib.annotations.ast.Property;
import me.bristermitten.mittenlib.annotations.config.ConfigProcessor;
import me.bristermitten.mittenlib.annotations.util.Nullity;
import me.bristermitten.mittenlib.config.Configuration;
import me.bristermitten.mittenlib.config.GeneratedConfig;
import me.bristermitten.mittenlib.config.Source;
import me.bristermitten.mittenlib.config.exception.ConfigLoadingErrors;

import javax.annotation.processing.Generated;
import javax.lang.model.element.Modifier;
import javax.lang.model.type.TypeMirror;
import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayDeque;
import java.util.List;
import java.util.Optional;

public class ConfigImplGenerator {

    private final AccessorGenerator accessorGenerator;
    private final ToStringGenerator toStringGenerator;
    private final EqualsHashCodeGenerator equalsHashCodeGenerator;
    private final ConfigurationClassNameGenerator configurationClassNameGenerator;
    private final ConfigNameCache configNameCache;
    private final MethodNames methodNames;

    @Inject
    public ConfigImplGenerator(
            AccessorGenerator accessorGenerator,
            ToStringGenerator toStringGenerator,
            EqualsHashCodeGenerator equalsHashCodeGenerator,
            ConfigurationClassNameGenerator configurationClassNameGenerator,
            ConfigNameCache configNameCache,
            MethodNames methodNames) {
        this.accessorGenerator = accessorGenerator;
        this.toStringGenerator = toStringGenerator;
        this.equalsHashCodeGenerator = equalsHashCodeGenerator;
        this.configurationClassNameGenerator = configurationClassNameGenerator;
        this.configNameCache = configNameCache;
        this.methodNames = methodNames;
    }

    private static void makeAbstractIfUnion(AbstractConfigStructure ast, TypeSpec.Builder source) {
        if (ast instanceof AbstractConfigStructure.Union) {
            source.addModifiers(Modifier.ABSTRACT);
        }
    }

    /**
     * Generates a JavaFile containing the implementation class for the given configuration structure.
     *
     * @param ast The abstract configuration structure to generate an implementation for
     * @return A JavaFile containing the generated implementation class
     */
    public JavaFile emit(AbstractConfigStructure ast) {
        ClassName configImplClassName =
                configurationClassNameGenerator.generateConfigurationClassName(ast.source().element());
        TypeSpec.Builder source = TypeSpec.classBuilder(configImplClassName);

        emitInto(ast, source);

        return JavaFile.builder(configImplClassName.packageName(), source.build()).build();
    }

    /**
     * Adds all necessary elements to the {@link TypeSpec.Builder} to create a complete implementation
     * class.
     *
     * @param ast    The abstract configuration structure to generate an implementation for
     * @param source The {@link TypeSpec.Builder} to add elements to
     */
    private void emitInto(AbstractConfigStructure ast, TypeSpec.Builder source) {
        ClassName configImplClassName =
                configurationClassNameGenerator.generateConfigurationClassName(ast.source().element());
        source.addModifiers(Modifier.PUBLIC);
        makeAbstractIfUnion(ast, source);
        addSourceElement(ast, source);
        addInheritance(ast, source);
        addInnerDefaultMethodImpl(source, ast);
        addGeneratedConfigAnnotations(ast, source);
        addNestedClassModifiers(ast, source);
        addProperties(ast, source);
        addSuperClassField(ast, source);
        accessorGenerator.createWithMethods(source, ast);
        addAllArgsConstructor(source, ast);
        addStandardObjectMethods(ast, configImplClassName, source);
        addChildClasses(ast, source);
    }

    /**
     * Adds the {@code CONFIG} static field to the class if a {@link Source} is defined.
     *
     * <p>Generates:
     *
     * <pre>{@code
     * public static final Configuration<MyConfig> CONFIG = new Configuration<>("config.yml", MyConfig.class);
     * }</pre>
     *
     * @param ast     the configuration structure, used to determine the public class name (e.g. {@code
     *                MyConfig}) and source (e.g. {@code "config.yml"})
     * @param builder the class builder
     */
    private void addSourceElement(AbstractConfigStructure ast, TypeSpec.Builder builder) {
        if (ast.settings().source() != null) {
            ClassName publicClassName = configurationClassNameGenerator.getPublicClassName(ast);
            ClassName implementationClassName =
                    configurationClassNameGenerator.translateConfigClassName(ast);

            FieldSpec.Builder configFieldBuilder =
                    FieldSpec.builder(
                                    ParameterizedTypeName.get(ClassName.get(Configuration.class), publicClassName),
                                    "CONFIG")
                            .addModifiers(Modifier.PUBLIC, Modifier.STATIC, Modifier.FINAL);

            configFieldBuilder.initializer(
                    "new $T<>($S, $T.class, $T.class)",
                    Configuration.class,
                    ast.settings().source().value(),
                    publicClassName,
                    implementationClassName);

            builder.addField(configFieldBuilder.build());
        }
    }

    /**
     * Adds inheritance information to the generated implementation class.
     *
     * <p>Generates:
     *
     * <pre>{@code
     * public class MyConfigImpl extends BaseConfigImpl implements MyConfig
     * }</pre>
     *
     * @param ast    the configuration structure, used to determine the parent class (e.g. {@code
     *               BaseConfigImpl}) and interfaces (e.g. {@code MyConfig})
     * @param source the class builder
     */
    private void addInheritance(AbstractConfigStructure ast, TypeSpec.Builder source) {
        if (ast.source() instanceof ConfigTypeSource.InterfaceConfigTypeSource) {
            source.addSuperinterface(ast.name());
        }
        if (ast.source() instanceof ConfigTypeSource.ClassConfigTypeSource classParent) {
            classParent
                    .parent()
                    .flatMap(configNameCache::lookupAST)
                    .ifPresent(
                            parent ->
                                    source.superclass(
                                            configurationClassNameGenerator.translateConfigClassName(parent)));
        }
    }

    /**
     * Adds {@link GeneratedConfig} and {@link Generated} annotations to the class.
     *
     * @param ast    the configuration structure
     * @param source the class builder
     */
    private void addGeneratedConfigAnnotations(AbstractConfigStructure ast, TypeSpec.Builder source) {
        final List<String> unserializableProperties =
                ast.properties().stream()
                        .filter(p -> !p.settings().hasDefaultValue() && !p.settings().isNullable())
                        .map(Property::name)
                        .toList();

        AnnotationSpec.Builder generatedConfigBuilder =
                AnnotationSpec.builder(GeneratedConfig.class)
                        .addMember("source", "$T.class", ast.name())
                        .addMember("isDynamicallyInitializable", "$L", ast.isDynamicallyInitializable());

        for (String property : unserializableProperties) {
            generatedConfigBuilder.addMember("uninitializableProperties", "$S", property);
        }

        source.addAnnotation(generatedConfigBuilder.build());

        source.addAnnotation(
                AnnotationSpec.builder(Generated.class)
                        .addMember("value", "$S", ConfigProcessor.class.getName())
                        .addMember("comments", "$S", "Generated by MittenLib Annotation Processor")
                        .addMember(
                                "date",
                                "$S",
                                ZonedDateTime.now(ZoneId.systemDefault()).format(DateTimeFormatter.ISO_INSTANT))
                        .build());
    }

    /**
     * Ensures nested classes are marked as {@code static}.
     *
     * @param ast    the configuration structure
     * @param source the class builder
     */
    private void addNestedClassModifiers(AbstractConfigStructure ast, TypeSpec.Builder source) {
        // if it's enclosed in a class, make sure it's a nested class rather than an inner class
        if (ast.enclosedIn() != null) {
            source.addModifiers(Modifier.STATIC);
        }
    }

    /**
     * Adds all properties as fields and accessors to the implementation class.
     *
     * @param ast    the configuration structure
     * @param source the class builder
     */
    private void addProperties(AbstractConfigStructure ast, TypeSpec.Builder source) {
        for (Property property : ast.properties()) {
            addProperty(property, source);
        }
    }

    /**
     * Adds {@code equals} and {@code hashCode} methods, plus {@code toString} if {@link
     * ASTSettings.ConfigASTSettings#generateToString()} is true
     *
     * @param ast                 the configuration structure
     * @param configImplClassName the name of the implementation class
     * @param source              the class builder
     */
    private void addStandardObjectMethods(
            AbstractConfigStructure ast, ClassName configImplClassName, TypeSpec.Builder source) {
        if (ast.settings().generateToString()) {
            var toString = toStringGenerator.generateToString(ast.properties(), configImplClassName);
            source.addMethod(toString);
        }

        source.addMethod(equalsHashCodeGenerator.generateEquals(configImplClassName, ast.properties()));
        source.addMethod(equalsHashCodeGenerator.generateHashCode(ast.properties()));
    }

    /**
     * Recursively adds implementation classes for enclosed configuration structures. For each
     * enclosed structure, we create a subclass builder and call {@link
     * #emitInto(AbstractConfigStructure, TypeSpec.Builder)} to add the subclass config values.
     *
     * @param ast    the parent configuration structure
     * @param source the parent class builder
     */
    private void addChildClasses(AbstractConfigStructure ast, TypeSpec.Builder source) {
        for (AbstractConfigStructure child : ast.enclosed()) {
            var childClassName = configurationClassNameGenerator.translateConfigClassName(child);
            TypeSpec.Builder childBuilder = TypeSpec.classBuilder(childClassName);
            emitInto(child, childBuilder);
            source.addType(childBuilder.build());
        }
    }

    /**
     * Adds a single property as a private final field and its corresponding getter.
     *
     * <p>Generates:
     *
     * <pre>{@code
     * private final String name;
     *
     * @Override
     * public String name() {
     *     return this.name;
     * }
     * }</pre>
     *
     * @param property the property to add, used for the field name and getter (e.g. {@code name})
     * @param source   the class builder
     */
    private void addProperty(Property property, TypeSpec.Builder source) {
        FieldSpec field =
                FieldSpec.builder(
                                configurationClassNameGenerator
                                        .publicPropertyClassName(property)
                                        .annotated(Nullity.getNullityAnnotationSpec(property)),
                                property.name(),
                                Modifier.FINAL,
                                Modifier.PRIVATE)
                        .build();

        source.addField(field);

        switch (property.source()) {
            case Property.PropertySource.FieldSource(var element) ->
                    accessorGenerator.createGetterMethod(source, element, field);
            case Property.PropertySource.MethodSource(var method) ->
                    accessorGenerator.createGetterMethodOverriding(source, method, field);
        }
    }

    private Optional<TypeMirror> getSuperClass(AbstractConfigStructure ast) {
        if (ast.source() instanceof ConfigTypeSource.ClassConfigTypeSource classParent) {
            return classParent.parent();
        }
        return Optional.empty();
    }

    private Optional<TypeMirror> getSuperClass(TypeMirror ast) {
        return configNameCache.lookupAST(ast).flatMap(this::getSuperClass);
    }

    /**
     * Adds an all-argument constructor to the implementation class.
     *
     * <p>Generates:
     *
     * <pre>{@code
     * public MyConfigImpl(String name, int age) {
     *     this.name = name;
     *     this.age = age;
     * }
     * }</pre>
     *
     * @param source the class builder
     * @param ast the configuration structure, used to determine constructor parameters (e.g. {@code
     *     name}, {@code age})
     */
    private void addAllArgsConstructor(TypeSpec.Builder source, AbstractConfigStructure ast) {
        MethodSpec.Builder constructor = MethodSpec.constructorBuilder().addModifiers(Modifier.PUBLIC);

        addSuperClassParameter(ast, constructor);
        addPropertyParameters(ast, constructor);

        source.addMethod(constructor.build());
    }

    /**
     * Adds a parameter to the constructor for the parent configuration class, if applicable.
     *
     * <p>Generates:
     *
     * <pre>{@code
     * public MyConfigImpl(BaseConfigImpl parent, String name) {
     *     super(parent.a(), parent.b());
     *     this.parent = parent;
     *     ... // see addPropertyParameters
     * }
     * }</pre>
     *
     * @param ast         the configuration structure, used to find the parent configuration
     * @param constructor the constructor builder to add the parameter and super call to
     */
    private void addSuperClassParameter(AbstractConfigStructure ast, MethodSpec.Builder constructor) {
        // when we have a super_class_
        // we accept an instance of it as a parent
        // and then call `super(parent.a(), parent.b(), ...)`
        var parentMirror = getSuperClass(ast);

        parentMirror.ifPresent(
                parent -> {
                    var parentConfig =
                            configNameCache
                                    .lookupAST(parent)
                                    .orElseThrow(
                                            () ->
                                                    new IllegalStateException(
                                                            "could not determine a config for parent class " + parent));
                    ClassName parentName =
                            configurationClassNameGenerator.translateConfigClassName(parentConfig);

                    String superParameterName = "parent";
                    constructor.addParameter(
                            ParameterSpec.builder(parentName, superParameterName, Modifier.FINAL).build());

                    List<String> parentParams =
                            buildSuperConstructorParams(parent, parentConfig, superParameterName);
                    constructor.addStatement("super($L)", String.join(", ", parentParams));
                    constructor.addStatement("this.parent = parent");
                });
    }

    /**
     * Adds a field to the implementation class to store the parent configuration instance.
     *
     * <p>Generates:
     *
     * <pre>{@code
     * private final BaseConfigImpl parent;
     * }</pre>
     *
     * @param ast the configuration structure, used to find the parent configuration
     * @param builder the class builder
     */
    private void addSuperClassField(AbstractConfigStructure ast, TypeSpec.Builder builder) {
        var parentMirror = getSuperClass(ast);

        parentMirror.ifPresent(
                parent -> {
                    var parentConfig =
                            configNameCache
                                    .lookupAST(parent)
                                    .orElseThrow(
                                            () ->
                                                    new IllegalStateException(
                                                            "could not determine a config for parent class " + parent));
                    ClassName parentName =
                            configurationClassNameGenerator.translateConfigClassName(parentConfig);
                    FieldSpec.Builder field =
                            FieldSpec.builder(parentName, "parent", Modifier.PRIVATE, Modifier.FINAL);

                    builder.addField(field.build());
                });
    }

    /**
     * Builds the list of parameters to be passed to the super constructor.
     *
     * <p>Generates:
     *
     * <pre>{@code
     * parent.a(), parent.b()
     * }</pre>
     *
     * @param parent             the type mirror of the parent class
     * @param parentConfig       the configuration structure of the parent
     * @param superParameterName the name of the parent parameter (e.g. {@code parent})
     * @return a list of code strings for the super constructor parameters
     */
    private List<String> buildSuperConstructorParams(
            TypeMirror parent, AbstractConfigStructure parentConfig, String superParameterName) {
        var parentParams =
                parentConfig.properties().stream()
                        .map(
                                variableElement ->
                                        superParameterName + "." + methodNames.safeMethodName(variableElement) + "()")
                        .toList();

        Optional<TypeMirror> parentParent = getSuperClass(parent);
        // when the superclass has a superclass, we need to also pass the superclass to the parent
        // constructor
        if (parentParent.isPresent()) {
            var newParentParams = new ArrayDeque<>(parentParams);
            newParentParams.addFirst(superParameterName);
            return List.copyOf(newParentParams);
        }

        return parentParams;
    }

    /**
     * Adds constructor parameters and initialization statements for all properties.
     *
     * <p>Generates:
     *
     * <pre>{@code
     * public MyConfigImpl(String name) {
     *     this.name = name;
     * }
     * }</pre>
     *
     * @param ast         the configuration structure
     * @param constructor the constructor builder
     */
    private void addPropertyParameters(AbstractConfigStructure ast, MethodSpec.Builder constructor) {
        for (Property property : ast.properties()) {
            ParameterSpec parameter = createPropertyParameter(property);
            constructor.addParameter(parameter);
            constructor.addStatement("this.$N = $N", property.name(), property.name());
        }
    }

    private ParameterSpec createPropertyParameter(Property property) {
        var nullityAnnotation = Nullity.getNullityAnnotation(property);
        ParameterSpec.Builder builder =
                ParameterSpec.builder(
                                configurationClassNameGenerator
                                        .publicPropertyClassName(property)
                                        .annotated(AnnotationSpec.builder(nullityAnnotation).build()),
                                property.name())
                        .addModifiers(Modifier.FINAL);

        return builder.build();
    }

    /**
     * Create a dummy interface named "DefaultMethodAccess" which implements the config interface, but
     * leaves any non-default methods empty. This essentially gives us an easy way to access the
     * default method of an interface. Any non-default method will be given an implementation that
     * throws {@link ConfigLoadingErrors#defaultValueProxyException(Class, String)}.
     *
     * <p>We only need to generate this interface if the given <code>ast</code> is interface and has
     * any properties with default values.
     *
     * @param typeSpecBuilder The {@link TypeSpec.Builder} to add elements to. The generated interface
     *     will be added here, if it should be generated.
     * @param ast the {@link AbstractConfigStructure} that we are generating a config from.
     * @return the {@link ClassName} of the generated <code>DefaultMethodAccess</code> interface,
     *     <i>if and only if</i> it was generated, otherwise an empty {@link Optional}.
     */
    private Optional<ClassName> addInnerDefaultMethodImpl(
            TypeSpec.Builder typeSpecBuilder, AbstractConfigStructure ast) {
        if (!(ast.source() instanceof ConfigTypeSource.InterfaceConfigTypeSource)) {
            return Optional.empty(); // nothing to do
        }

        boolean hasAnyDefaultValue =
                ast.properties().stream().anyMatch(property -> property.settings().hasDefaultValue());
        // if no properties have default values, there's nothing to do
        if (!hasAnyDefaultValue) {
            return Optional.empty();
        }

        ClassName concreteConfigClassName =
                configurationClassNameGenerator.getConcreteConfigClassName(ast);
        var innerName = configurationClassNameGenerator.getDefaultMethodAccessClassName(ast);

        var innerBuilder = TypeSpec.classBuilder(innerName);
        innerBuilder.addModifiers(Modifier.PUBLIC, Modifier.STATIC);
        innerBuilder.addSuperinterface(ast.name());

        for (Property property : ast.properties()) {
            if (property.settings().hasDefaultValue()) {
                continue;
            }

            innerBuilder.addMethod(
                    MethodSpec.methodBuilder(property.name())
                            .addModifiers(Modifier.PUBLIC)
                            .addAnnotation(Override.class)
                            .returns(configurationClassNameGenerator.publicPropertyClassName(property))
              .addStatement(
                  "throw $T.defaultValueProxyException($T.class, $S)",
                  ConfigLoadingErrors.class,
                  concreteConfigClassName,
                  property.name())
              .build());
    }

    typeSpecBuilder.addType(innerBuilder.build());
    return Optional.of(innerName);
  }
}

File: annotation-processor/src/main/scala/me/bristermitten/mittenlib/annotations/compile/ConfigImplGenerator.scala
package me.bristermitten.mittenlib.annotations.compile

import com.google.inject.Inject
import com.palantir.javapoet.*
import java.time.ZoneId
import java.time.ZonedDateTime
import java.time.format.DateTimeFormatter
import java.util.{ArrayDeque, List => JList, Optional => JOptional}
import javax.annotation.processing.Generated
import javax.lang.model.element.{Modifier, TypeElement, Element}
import javax.lang.model.`type`.TypeMirror
import me.bristermitten.mittenlib.annotations.domain.*
import me.bristermitten.mittenlib.annotations.ast.AbstractConfigStructure
import me.bristermitten.mittenlib.annotations.config.ConfigProcessor
import me.bristermitten.mittenlib.annotations.util.ConfigStructureAnalysis
import me.bristermitten.mittenlib.annotations.util.ElementsFinder
import me.bristermitten.mittenlib.config.Configuration
import me.bristermitten.mittenlib.config.GeneratedConfig
import me.bristermitten.mittenlib.config.Source
import me.bristermitten.mittenlib.config.exception.ConfigLoadingErrors
import javax.annotation.processing.ProcessingEnvironment
import scala.jdk.CollectionConverters.*
import scala.jdk.OptionConverters.*

class ConfigImplGenerator @Inject() (
    private val accessorGenerator: AccessorGenerator,
    private val toStringGenerator: ToStringGenerator,
    private val equalsHashCodeGenerator: EqualsHashCodeGenerator,
    private val configurationClassNameGenerator: ConfigurationClassNameGenerator,
    private val configNameCache: ConfigNameCache,
    private val methodNames: MethodNames,
    private val configStructureAnalysis: ConfigStructureAnalysis,
    private val elementsFinder: ElementsFinder,
    processingEnv: ProcessingEnvironment
):
  private val elements = processingEnv.getElementUtils

  /** Generates a JavaFile containing the implementation class for the given
    * configuration structure.
    *
    * @param ast
    *   The abstract configuration structure to generate an implementation for
    * @return
    *   A JavaFile containing the generated implementation class
    */
  def emit(ast: ConfigStructure): JavaFile =
    val dtoType = elements.getTypeElement(ast.name.canonicalName())
    val configImplClassName =
      configurationClassNameGenerator.generateConfigurationClassName(dtoType)
    val source = TypeSpec
      .classBuilder(configImplClassName)
      .addJavadoc(
        """Generated data implementation of {@link $T}.
          |""".stripMargin,
        dtoType
      )

    emitInto(ast, source)

    JavaFile
      .builder(configImplClassName.packageName(), source.build())
      .skipJavaLangImports(true)
      .build()

  /** Adds all necessary elements to the {@link TypeSpec.Builder} to create a
    * complete implementation class.
    */
  private def emitInto(
      ast: ConfigStructure,
      source: TypeSpec.Builder
  ): Unit =
    val dtoType = elements.getTypeElement(ast.name.canonicalName())
    val configImplClassName =
      configurationClassNameGenerator.generateConfigurationClassName(dtoType)
    source.addModifiers(Modifier.PUBLIC)
    makeAbstractIfUnion(ast, source)
    addSourceElement(ast, source)
    addInheritance(ast, source)
    addInnerDefaultMethodImpl(source, ast)
    addGeneratedConfigAnnotations(ast, source)
    addNestedClassModifiers(ast, source)
    addProperties(ast, source)
    addSuperClassField(ast, source)
    configNameCache
      .lookupAST(ast.name)
      .toScala
      .foreach(rawAst => accessorGenerator.createWithMethods(source, rawAst))
    addAllArgsConstructor(source, ast)
    if (
      configStructureAnalysis.isDynamicallyInitializable(ast) && !ast
        .isInstanceOf[ConfigStructure.Union]
    ) {
      if (ast.properties.nonEmpty || getParentConfig(ast).isDefined) {
        addNoArgsConstructor(source, ast)
      }
    }
    addStandardObjectMethods(ast, configImplClassName, source)
    addChildClasses(ast, source)

  private def makeAbstractIfUnion(
      ast: ConfigStructure,
      source: TypeSpec.Builder
  ): Unit =
    ast match {
      case _: ConfigStructure.Union =>
        source.addModifiers(Modifier.ABSTRACT)
      case _ =>
    }

  /** Adds the {@code CONFIG} static field to the class if a {@link Source} is
    * defined.
    */
  private def addSourceElement(
      ast: ConfigStructure,
      builder: TypeSpec.Builder
  ): Unit =
    ast.settings.source.foreach { sourceVal =>
      val dtoType = elements.getTypeElement(ast.name.canonicalName())
      val publicClassName =
        configurationClassNameGenerator.getPublicClassName(ast)
      val implementationClassName =
        configurationClassNameGenerator.translateConfigClassName(ast)

      val configFieldBuilder = FieldSpec
        .builder(
          ParameterizedTypeName
            .get(ClassName.get(classOf[Configuration[?]]), publicClassName),
          "CONFIG"
        )
        .addModifiers(Modifier.PUBLIC, Modifier.STATIC, Modifier.FINAL)

      configFieldBuilder.initializer(
        "new $T<>($S, $T.class, $T.class)",
        classOf[Configuration[?]],
        sourceVal,
        publicClassName,
        implementationClassName
      )

      builder.addField(configFieldBuilder.build())
    }

  /** Adds inheritance information to the generated implementation class.
    */
  private def addInheritance(
      ast: ConfigStructure,
      source: TypeSpec.Builder
  ): Unit =
    ast match {
      case ConfigStructure.Atomic(name, isInterface, parentClass, _, _, _) =>
        if (isInterface) {
          source.addSuperinterface(name)
        } else {
          for {
            parentType <- parentClass
            parentAst <- configNameCache.lookupAST(parentType).toScala
          } {
            source.superclass(
              configurationClassNameGenerator.translateConfigClassName(
                parentAst
              )
            )
          }
        }
      case ConfigStructure.Intersection(name, _, _, _, _) =>
        source.addSuperinterface(name)
      case ConfigStructure.Union(name, _, _, _, _) =>
        source.addSuperinterface(name)
    }

  /** Adds {@link GeneratedConfig} and {@link Generated} annotations to the
    * class.
    */
  private def addGeneratedConfigAnnotations(
      ast: ConfigStructure,
      source: TypeSpec.Builder
  ): Unit =
    val uninitializableProperties = ast.properties
      .filter(p =>
        !p.hasDefault
          && !p.isNullable
          && !configStructureAnalysis.isTypeInitializable(p.propertyType)
      )
      .map(_.name)

    val generatedConfigBuilder = AnnotationSpec
      .builder(classOf[GeneratedConfig])
      .addMember("source", "$T.class", ast.name)
      .addMember(
        "isDynamicallyInitializable",
        "$L",
        Boolean.box(configStructureAnalysis.isDynamicallyInitializable(ast))
      )

    for (property <- uninitializableProperties) {
      generatedConfigBuilder.addMember(
        "uninitializableProperties",
        "$S",
        property
      )
    }

    source.addAnnotation(generatedConfigBuilder.build())
    source.addAnnotation(GeneratorUtil.generatedAnnotation(true))

  /** Ensures nested classes are marked as {@code static} .
    */
  private def addNestedClassModifiers(
      ast: ConfigStructure,
      source: TypeSpec.Builder
  ): Unit =
    if (ast.name.enclosingClassName() != null) {
      source.addModifiers(Modifier.STATIC)
    }

  /** Adds all properties as fields and accessors to the implementation class.
    */
  private def addProperties(
      ast: ConfigStructure,
      source: TypeSpec.Builder
  ): Unit =
    val dtoType = elements.getTypeElement(ast.name.canonicalName())
    for (property <- ast.properties) {
      addProperty(ast, property, dtoType, source)
    }

  /** Adds {@code equals} and {@code hashCode} methods, and {@code toString}
    */
  private def addStandardObjectMethods(
      ast: ConfigStructure,
      configImplClassName: ClassName,
      source: TypeSpec.Builder
  ): Unit =
    val propertiesJavaList =
      ast.properties.asJava // For toString/equals/hashCode generators
    // toString, equals, hashCode are always generated now
    val toString = toStringGenerator.generateToStringDomain(
      propertiesJavaList,
      configImplClassName
    )
    source.addMethod(toString)

    source.addMethod(
      equalsHashCodeGenerator.generateEqualsDomain(
        configImplClassName,
        propertiesJavaList
      )
    )
    source.addMethod(
      equalsHashCodeGenerator.generateHashCodeDomain(propertiesJavaList)
    )

  /** Recursively adds implementation classes for enclosed configuration
    * structures.
    */
  private def addChildClasses(
      ast: ConfigStructure,
      source: TypeSpec.Builder
  ): Unit =
    for (child <- ast.enclosed) {
      val childClassName =
        configurationClassNameGenerator.translateConfigClassName(child)
      val childBuilder = TypeSpec.classBuilder(childClassName)
      emitInto(child, childBuilder)
      source.addType(childBuilder.build())
    }

  private def getAnnotatedPropertyType(property: Property): TypeName =
    val nullityAnnotation =
      if (property.isNullable) classOf[org.jspecify.annotations.Nullable]
      else classOf[org.jspecify.annotations.NonNull]
    configurationClassNameGenerator
      .publicPropertyClassName(property.typeMirror)
      .annotated(AnnotationSpec.builder(nullityAnnotation).build())

  /** Adds a single property as a private final field and its corresponding
    * getter.
    */
  private def addProperty(
      ast: ConfigStructure,
      property: Property,
      dtoType: TypeElement,
      source: TypeSpec.Builder
  ): Unit =
    val field = FieldSpec
      .builder(
        getAnnotatedPropertyType(property),
        property.name,
        Modifier.FINAL,
        Modifier.PRIVATE
      )
      .build()

    source.addField(field)

    val isInterface = ast match {
      case ConfigStructure.Atomic(_, isInterface, _, _, _, _) => isInterface
      case _                                                  => true
    }

    if (isInterface) {
      val methodOpt = elementsFinder
        .getPropertyMethods(dtoType)
        .find(_.getSimpleName.toString == property.name)
      methodOpt.foreach { method =>
        accessorGenerator.createGetterMethodOverriding(source, method, field)
      }
    } else {
      val fieldOpt = elementsFinder
        .getApplicableVariableElements(dtoType)
        .find(_.getSimpleName.toString == property.name)
      fieldOpt.foreach { f =>
        accessorGenerator.createGetterMethod(source, f, field)
      }
    }

  private def getSuperClass(ast: ConfigStructure): Option[ClassName] =
    ast match {
      case ConfigStructure.Atomic(_, _, parentClass, _, _, _) =>
        parentClass
      case _ =>
        None
    }

  /** Adds an all-argument constructor to the implementation class.
    */
  private def addAllArgsConstructor(
      source: TypeSpec.Builder,
      ast: ConfigStructure
  ): Unit =
    val constructor = MethodSpec
      .constructorBuilder()
      .addJavadoc(
        "Constructs a new implementation instance with all properties populated.\n"
      )
      .addModifiers(Modifier.PUBLIC)

    addSuperClassParameter(ast, constructor)
    addPropertyParameters(ast, constructor)

    source.addMethod(constructor.build())

  private def getParentConfig(
      ast: ConfigStructure
  ): Option[(ClassName, AbstractConfigStructure)] =
    getSuperClass(ast).map { parent =>
      val parentConfig = configNameCache
        .lookupAST(parent)
        .toScala
        .getOrElse(
          throw new IllegalStateException(
            "could not determine a config for parent class " + parent
          )
        )
      (parent, parentConfig)
    }

  /** Adds a parameter to the constructor for the parent configuration class, if
    * applicable.
    */
  private def addSuperClassParameter(
      ast: ConfigStructure,
      constructor: MethodSpec.Builder
  ): Unit =
    getParentConfig(ast).foreach { case (parent, parentConfig) =>
      val parentName =
        configurationClassNameGenerator.translateConfigClassName(parentConfig)
      val superParameterName = "parent"
      constructor.addParameter(
        ParameterSpec
          .builder(parentName, superParameterName, Modifier.FINAL)
          .build()
      )

      val parentParams =
        buildSuperConstructorParams(parent, parentConfig, superParameterName)
      constructor.addStatement("super($L)", parentParams.mkString(", "))
      constructor.addStatement("this.parent = parent")
    }

  /** Adds a field to the implementation class to store the parent configuration
    * instance.
    */
  private def addSuperClassField(
      ast: ConfigStructure,
      builder: TypeSpec.Builder
  ): Unit =
    getParentConfig(ast).foreach { case (_, parentConfig) =>
      val parentName =
        configurationClassNameGenerator.translateConfigClassName(parentConfig)
      val field = FieldSpec.builder(
        parentName,
        "parent",
        Modifier.PRIVATE,
        Modifier.FINAL
      )
      builder.addField(field.build())
    }

  /** Builds the list of parameters to be passed to the super constructor.
    */
  private def buildSuperConstructorParams(
      parent: ClassName,
      parentConfig: AbstractConfigStructure,
      superParameterName: String
  ): List[String] =
    val parentParams = parentConfig
      .properties()
      .asScala
      .toList
      .map(p => superParameterName + "." + methodNames.safeMethodName(p) + "()")

    val grandParentExists = parentConfig.source() match {
      case c: me.bristermitten.mittenlib.annotations.ast.ConfigTypeSource.ClassConfigTypeSource =>
        c.parentField.isPresent
      case _ => false
    }

    if (grandParentExists) {
      superParameterName :: parentParams
    } else {
      parentParams
    }

  /** Adds constructor parameters and initialization statements for all
    * properties.
    */
  private def addPropertyParameters(
      ast: ConfigStructure,
      constructor: MethodSpec.Builder
  ): Unit =
    for (property <- ast.properties) {
      val parameter = createPropertyParameter(property)
      constructor.addParameter(parameter)
      constructor.addStatement("this.$N = $N", property.name, property.name)
    }

  private def createPropertyParameter(property: Property): ParameterSpec =
    val builder = ParameterSpec
      .builder(
        getAnnotatedPropertyType(property),
        property.name
      )
      .addModifiers(Modifier.FINAL)

    builder.build()

  private def addNoArgsConstructor(
      source: TypeSpec.Builder,
      ast: ConfigStructure
  ): Unit =
    val constructor = MethodSpec
      .constructorBuilder()
      .addJavadoc(
        "Constructs a new implementation instance with default values.\n"
      )
      .addModifiers(Modifier.PUBLIC)

    val parentOpt = getParentConfig(ast)
    if (parentOpt.isDefined) {
      constructor.addStatement("super()")
      constructor.addStatement("this.parent = null")
    }

    val hasAnyDefaultValue =
      ast.properties.exists(configStructureAnalysis.hasDefaultOrIsInitializable)
    if (
      hasAnyDefaultValue && (ast match {
        case atomic: ConfigStructure.Atomic  => atomic.isInterface
        case _: ConfigStructure.Intersection => true
        case _: ConfigStructure.Union        => false
      })
    ) {
      val innerName =
        configurationClassNameGenerator.getDefaultMethodAccessClassName(ast)
      constructor.addStatement(
        "$T defaultAccess = new $T()",
        innerName,
        innerName
      )
      for (property <- ast.properties) {
        constructor.addStatement(
          "this.$N = defaultAccess.$N()",
          property.name,
          property.name
        )
      }
    } else {
      for (property <- ast.properties) {
        val typeName = configurationClassNameGenerator.publicPropertyClassName(
          property.typeMirror
        )
        if (typeName.isPrimitive) {
          if (typeName == TypeName.BOOLEAN) {
            constructor.addStatement("this.$N = false", property.name)
          } else {
            constructor.addStatement("this.$N = 0", property.name)
          }
        } else {
          constructor.addStatement("this.$N = null", property.name)
        }
      }
    }

    source.addMethod(constructor.build())

  /** Create a dummy class/interface named "DefaultMethodAccess".
    */
  private def addInnerDefaultMethodImpl(
      typeSpecBuilder: TypeSpec.Builder,
      ast: ConfigStructure
  ): JOptional[ClassName] =
    val isInterfaceWithDefaults = ast match {
      case ConfigStructure.Atomic(_, true, _, _, properties, _) =>
        Some(properties)
      case ConfigStructure.Intersection(_, _, _, properties, _) =>
        Some(properties)
      case _ => None
    }
    isInterfaceWithDefaults match {
      case None             => JOptional.empty()
      case Some(properties) =>
        val hasAnyDefaultValue =
          properties.exists(configStructureAnalysis.hasDefaultOrIsInitializable)
        if (!hasAnyDefaultValue) {
          JOptional.empty()
        } else {
          val concreteConfigClassName =
            configurationClassNameGenerator.getConcreteConfigClassName(ast)
          val innerName =
            configurationClassNameGenerator.getDefaultMethodAccessClassName(ast)

          val innerBuilder = TypeSpec.classBuilder(innerName)
          innerBuilder.addModifiers(Modifier.PUBLIC, Modifier.STATIC)
          innerBuilder.addSuperinterface(ast.name)

          for (
            property <- properties
            if !property.hasDefault
          ) {
            val methodBuilder = MethodSpec
              .methodBuilder(property.name)
              .addModifiers(Modifier.PUBLIC)
              .addAnnotation(classOf[Override])
              .returns(
                configurationClassNameGenerator
                  .publicPropertyClassName(property.typeMirror)
              )

            if (
              configStructureAnalysis.isTypeInitializable(property.propertyType)
            ) {
              property.propertyType match {
                case PropertyType.ConfigProperty(className, _) =>
                  val concreteType =
                    configurationClassNameGenerator.translateConfigClassName(
                      className
                    )
                  methodBuilder.addStatement("return new $T()", concreteType)
                case PropertyType.OptionalProperty(_) =>
                  methodBuilder.addStatement(
                    "return $T.empty()",
                    classOf[java.util.Optional[?]]
                  )
                case other =>
                  throw new IllegalStateException(
                    s"Type marked initializable but has no default generator: $other"
                  )
              }
            } else if (property.isNullable) {
              methodBuilder.addStatement("return null")
            } else {
              methodBuilder.addStatement(
                "throw $T.defaultValueProxyException($T.class, $S)",
                classOf[ConfigLoadingErrors],
                concreteConfigClassName,
                property.name
              )
            }
            innerBuilder.addMethod(methodBuilder.build())
          }

          typeSpecBuilder.addType(innerBuilder.build())
          JOptional.of(innerName)
        }
    }
